tket/src/WeightSubgrMono/Searching/NodeListTraversal.cpp

```diff
@@ -0,0 +1,289 @@
+
+std::set<VertexWSM> NodeListTraversal::get_used_target_vertices() const {
+  std::set<VertexWSM> target_vertices;
+
+  // Examine all PV.
+  for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+    const auto& domain_data = m_raw_data.domains_data[pv];
+
+    // Examine Dom(PV) at all levels.
+    for (unsigned level = 0; level <= domain_data.entries_back_index; ++level) {
+      const auto& node_index = domain_data.entries[level].node_level;
+      // Nogood nodes are ignored.
+      // So, check that at least one node with this domain
+      // is not a nogood; otherwise skip.
+      if (m_raw_data.nodes_data[node_index].nogood &&
+          level < domain_data.entries_back_index) {
+        const auto& next_node_index = domain_data.entries[level + 1].node_level;
+        bool some_node_is_good = false;
+        for (auto ii = node_index; ii < next_node_index; ++ii) {
+          if (!m_raw_data.nodes_data[ii].nogood) {
+            some_node_is_good = true;
+            break;
+          }
+        }
+        if (!some_node_is_good) {
+          // Skip.
+          continue;
+        }
+      }
+      const auto& domain = domain_data.entries[level].domain;
+      if (target_vertices.empty()) {
+        target_vertices = domain;
+      } else {
+        for (VertexWSM tv : domain) {
+          target_vertices.insert(tv);
+        }
+      }
+    }
+  }
+  return target_vertices;
+}
+
+bool NodeListTraversal::move_up() {
+  while (m_raw_data.current_node_level > 0) {
+    --m_raw_data.current_node_level;
+    if (m_raw_data.nodes_data[m_raw_data.current_node_level].nogood) {
+      continue;
+    }
+    for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+      auto& domain_data = m_raw_data.domains_data[pv];
+      while (domain_data.entries[domain_data.entries_back_index].node_level >
+             m_raw_data.current_node_level) {
+        // We've moved above the level of "junk data",
+        // so shrink the back index.
+        TKET_ASSERT(domain_data.entries_back_index > 0);
+        --domain_data.entries_back_index;
+      }
+    }
+    return true;
+  }
+  // We've hit the top!
+  return false;
+}
+
+// Returns false if the current node becomes invalid.
+// Also handles the new assignment PV->y, if necessary,
+// for the CURRENT node (not new node).
+static bool when_moving_down_check_current_domain_size(
+    const std::set<VertexWSM>& existing_domain, NodesRawData& raw_data,
+    VertexWSM p_vertex, VertexWSM t_vertex) {
+  switch (existing_domain.size()) {
+    case 0:
+      TKET_ASSERT(false);
+      return false;
+    case 1:
+      // Already Dom(PV) = {TV}, so it will become a nogood.
+      raw_data.get_current_node_nonconst().nogood = true;
+      return false;
+    case 2: {
+      // Dom(PV) = {TV, y}, so it will become Dom(PV) = {y}
+      // at the current level, so we need a new assignment PV->y.
+      VertexWSM tv_other = *existing_domain.cbegin();
```

Comment:
We are moving DOWN from this node; therefore the next time we come back UP to this node, we will not repeat this choice.

The simplest way to ensure this is to remove the choice TV from Domain(PV).

However, since Domain(PV) = {TV, y} has size 2, AFTER we remove TV it will become Domain(PV) = {y} of size 1.

This is perfectly fine; we're leaving it here in this node, READY for when we move back up to it in future; HOWEVER domains of size 1 must be treated differently (they create a new assignment PV->y) and so we ADD this assignment to the list of assignments in THIS node needing to be processed.

Of course we won't actually process it until we've moved back up to this node at some point in the future (after moving down, which we are about to do), but it leave it sitting there, ready.

This point also caused me a lot of confusion before; we should NOT think of nodes as the actual history of previous nodes we passed through. Instead, we should think of nodes as being the next state to make a move, when it comes time to backtrack (move up) to it. (It's really just "for convenience" that the nodes are ordered and visited in this way; it's the simplest choice, but not the only possibility).